### PR TITLE
DM-28035: Backfill tool for DBB endpoint manager (Stage 1)

### PR DESCRIPTION
**Pull request by mxk62**
Added a backfill tool allowing one to populate endpoint manager's table with entries related to files which were already in the storage area before its initial deployment.

Note that this tool does *not* check if a file is ingested to the Butler dataset repository managed by the endpoint manager nor it tries to ingest the file if it isn't. It merely makes the endpoint manager "aware of" these files.

### CODE REVIEW

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 1]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+        daemon: false
+        file_status: RERUN
 
 When run in non-daemonic mode, the Ingester will quit after processing all the
 files with ``RERUN`` status.
+
+Backfill database records
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Most likely some files were already present in the storage area and ingested to
```
Comment (MichelleGower) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: Most likely -> In the case where some files ...

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 2]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+The statuses of all entries added by the tool to the ``file_events`` table will
+be set to BACKFILL.
+
+The backfill tool will exit when it processes all the files in the provided
+source directory (or directories).
+
+
+.. warning::
+
+   At the moment, the backfill tool does not verify if any of the files found
```
Comment (MichelleGower) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: Remove "At the moment"

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        required = {"session", "sources", "storage", "tablenames"}
+        missing = required - set(config)
+        if missing:
+            msg = f"invalid configuration: {', '.join(missing)} not provided"
+            logger.error(msg)
+            raise ValueError(msg)
+
+        self.session = config["session"]
+
+        required = {"event", "file"}
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:78]: Move shared code to function or at least comment.  It took me a minute to figure out why this second block of code existed because it looks so much similar to the previous block.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+
+logger = logging.getLogger(__name__)
+
+
+@click.group()
+def backfill():
+    """Populate the database with entries for historical files.
+    """
+    pass
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:38]: I don't know enough about click groups.  Just verifying that this method is necessary since it only contains pass or is the pass line needed?

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        search = config["search"]
+        self.search_opts = dict(blacklist=search.get("blacklist", None))
+
+    def run(self):
+        """Start the backfill process.
+        """
+        acc = dict(failure=0, success=0)
+        sources = [glob(os.path.join(self.storage, s)) for s in self.sources]
+        for source in chain(*sources):
+            logger.debug(f"{source}: setting as file source")
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:114]: Use lazy % formatting in logging functions (https://developer.lsst.io/stack/logging.html#basic-usage-in-python but it seems to be LSST logging specific.)  If want to make a separate ticket to make this change across entire code, feel free to postpone.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+            paths = scan(source, **self.search_opts) if os.path.isdir(source) \
+                else [source]
+            for path in paths:
+                abspath = os.path.join(source, path)
+                relpath = os.path.relpath(abspath, start=self.storage)
+                dirname, filename = os.path.split(relpath)
+                logger.debug(f"{relpath}: starting processing")
+
+                logger.debug(f"checking if already tracked")
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: f-string without any variables (i.e., remove the _f_)  Multiple lines are this way.

File: python/lsst/dbb/buffmngrs/endpoint/search.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    -------
+    gen : generator
+        The original, unchanged generator.
+    """
+    search_methods[gen.__name__] = gen
+    return gen
+
+
+@register
+def scan(directory, blacklist=None, **kwargs):
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/search.py:60]: kwargs not used (or if really want to keep, missing in docstring)

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+"""Command line interface for the Backfill component.
+"""
+import click
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py]: click is a third-party import and should go later than standard imports (e.g. imports of importlib, inspect, and logging).

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        msg = "runtime environment and configuration:\n\n"
+        msg += dump_env()
+        msg += "\n"
+        msg += dump_config(configuration)
+        logger.info(msg)
+
+    logger.info("setting up database connection...")
+    config = configuration["database"]
+
+    module = importlib.import_module("sqlalchemy.pool")
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py]: Should this be using LSST's doImport?

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 10]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
 files with ``RERUN`` status.
+
+Backfill database records
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Most likely some files were already present in the storage area and ingested to
+a Butler dataset repository before the deployment of the DBB endpoint buffer
+manager.  
+
+Use ``backfill`` tool to populate the endpoint manager's database tables with
```
Comment (MichelleGower) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: tool -> command or subcommand as went with subcommands instead of separate script?

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+from ..utils import get_checksum
+
+
+__all__ = ["Backfill"]
+
+
+logger = logging.getLogger(__name__)
+
+
+class Backfill(object):
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: object isn't needed.  In Python 3 inheritance from object is implicit.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+@backfill.command()
+@click.option("--dump/--no-dump", default=True,
+              help="Log runtime environment and configuration "
+                   "(ignored if severity is set to WARNING and higher).")
+@click.option("--validate/--no-validate", default=False,
+              help="Validate configuration before starting the service.")
+@click.argument("filename", type=click.Path(exists=True))
+def start(filename, dump, validate):
+    """Starts a finder using a configuration from FILENAME.
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py]: "Starts a **finder**"?  Is this a copy/paste mistake or have I misunderstood something?

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+    """
+    with open(filename) as f:
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:51]: A bunch of the start function is the same across all three */commands.py files.  The shared code should be moved to function(s).  There are blocks of code that could be smaller functions (e.g., setting up the sqlalchemy pool) or might be able keep as reusable larger function.  Again, ok if want to postpone to a new ticket.

File: python/lsst/dbb/buffmngrs/endpoint/validation.py [hunk 14]
```diff
@@ -286,3 +289,93 @@
     - database
     - ingester
 """
+
+backfill = """
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/validation.py]: Constants should be UPPER_CASE?

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 15]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+entires regarding these files.
```
Comment (MichelleGower) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: typo: entires -> entries?

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 16]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+          - "2020-12-31"
+        search:
+          blacklist:
+            - \.txt$
+
+   .. note::
+
+      You can specify multiple sources.  The source entry can be either a file
+      or a directory.  You can use unix style pathname pattern expansion
+      when specyfing source's name (tilde expansion is not supported though).
```
Comment (MichelleGower) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: typo: specyfing -> specifying

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 17]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+      All sources are assumed to be relative with regard to the storage area.
+
+   .. note::
+
+      To prevent Python YAML parser from accidental conversion of strings like
+      ``2021-01-01`` into a ``datatime`` objects, enclose them in quotes.
```
Comment (MichelleGower) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: typo: datatime -> datetime

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+                    logger.error(f"{relpath}: cannot create database entries: "
+                                 f"{ex}")
+                    acc["failure"] += 1
+                else:
+                    logger.debug(f"{relpath}: processing completed")
+                    acc["success"] += 1
+        total = sum(x for x in acc.values())
+        if total == 0:
+            logger.warning("no backfill attempts were made")
+        if acc["failure"] != 0:
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: Not sure why this isn't an elif (i.e., if no backfill **attempts** were made can one have failures or successes?).

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 19]
```diff
@@ -0,0 +1,190 @@
...
+            raise ValueError(msg)
+        self.sources = config["sources"]
+
+        search = config["search"]
+        self.search_opts = dict(blacklist=search.get("blacklist", None))
+
+    def run(self):
+        """Start the backfill process.
+        """
+        acc = dict(failure=0, success=0)
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: Seems to be keeping track of counts as opposed to the more generic accumulating values.  Code would be easier to read if variable name was some form of the word counts.

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 20]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+        search:
+          blacklist:
+            - \.txt$
+
+   .. note::
+
+      You can specify multiple sources.  The source entry can be either a file
+      or a directory.  You can use unix style pathname pattern expansion
+      when specyfing source's name (tilde expansion is not supported though).
+      All sources are assumed to be relative with regard to the storage area.
```
Comment (MichelleGower) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: If abs paths that include the storage path are given does it work?  If it breaks, change the wording to "must be" and consider checking not abs paths.  If it works and are ok with someone using it, also change the wording.  Code definitely should not support absolute paths that do not include the storage path.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+        sources = [glob(os.path.join(self.storage, s)) for s in self.sources]
+        for source in chain(*sources):
+            logger.debug(f"{source}: setting as file source")
+            if not os.path.exists(source):
+                logger.warning(f"{source}: no such file or directory")
+                acc["failure"] += 1
+                continue
+
+            paths = scan(source, **self.search_opts) if os.path.isdir(source) \
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:123]: This section of code is really hard to follow.   Please add more comments.  Also feels like there is a bunch of path manipulations that could be cleaned up (e.g., adding, removing, adding, removing).  Seems like there should be some way of removing the self.storage from the source and then only add the additional bits to the path in the loop.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py (hunk 19 above)
Comment (MichelleGower) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: Also after reading farther down the code, failure covers many different failure mechanisms including already tracked.  It would be helpful if these counts were separated into more categories, especially those that are unusual internal issues (sqlalchemy outage, file disappeared in the middle of processing, etc) vs already in DB (which should just be an informational count so one can rerun backfill on the same directories without it reporting already tracked as failures).

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 22]
```diff
@@ -0,0 +1,190 @@
...
+                else:
+                    logger.debug(f"{relpath}: processing completed")
+                    acc["success"] += 1
+        total = sum(x for x in acc.values())
+        if total == 0:
+            logger.warning("no backfill attempts were made")
+        if acc["failure"] != 0:
+            logger.warning("some backfill attempts failed")
+        else:
+            logger.info("all found files were backfilled successfully")
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: Since the code has the count of successes why not print it?

Comment (MichelleGower) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: Also referring back to the earlier comment about "already tracked" shouldn't be an error (i.e., I should be able to run and rerun backfill on the same directory and it should only be errors if it can't backfill files that need to be backfilled), a count of the "already tracked" could be printed here as info (or warning).

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py [hunk 23]
```diff
@@ -0,0 +1,139 @@
...
+    # settings from relevant section of the global configuration, but new
+    # settings may be added, already existing ones may be altered.
+    backfill_config = dict(config)
+
+    backfill_config["session"] = session
+    backfill_config["tablenames"] = tablenames
+
+    logger.info("starting Backfill...")
+    component = Backfill(backfill_config)
+    component.run()
```
Comment (MichelleGower) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:84]: Not exactly sure what line to add this comment with, but if there are backfill failures/errors (not counting "already tracked" as errors), shouldn't endmgr exit with a non-zero exit code?

**Review by MichelleGower — APPROVED** [review]
Comments varying from typos to reorganize shared code.  A few of this I noted could be postponed to later ticket(s).  If so, please go ahead and make the new tickets.  There was one section of code that I noted was difficult for me to figure out.   While trying to figure it out, I was able to set up and run some tests of my own.  This means I could understand the instructions and the test cases worked.  Feel free to merge the code after fixing (or creating the new tickets).

File: python/lsst/dbb/buffmngrs/endpoint/search.py (hunk 7 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/search.py:60]: Yes, I want to keep it to ensure that search methods can be used interchangeably. I will document it in the docstring

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py (hunk 9 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py]: Nothing in the _framework_ depends on LSST Stack software. Only GenX Butler ingest plugins use LSST code. I'd prefer to keep it this way in case plugins will end up in a separate package.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py (hunk 4 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:38]: As far as my understanding of Click goes, yes, it is necessary to nest commands.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py (hunk 12 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py]: Yes, it's a copy/paste mistake. Will fix.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py (hunk 5 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:114]: None of the manager's commands use lazy formatting at the moment. I'd like to keep things consistent so I'll make a separate ticket for that work.

As a side note, all f-strings used by log messages include basic Python data structures only so the overhead resulting from eager formatting should be negligible.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py (hunk 19 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: You're totally right. I will:
1. change "acc" to "counts",
2. introduce new counter "tracked",
3. decrease the severity of the log message about already track file from WARNING to INFO.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py (hunk 22 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: > Since the code has the count of successes why not print it?

No particular reason, I just assumed no one would care how many files were backfilled successfully _if_ there were no failures (of any kind).

However, in the context your earlier suggestion about distinguishing between different types of failed backfills (DB errors vs. already tracked), now it will make sense to presents these statistics. I will make the necessary changes.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py (hunk 23 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:84]: I'll make the code to raise a RuntimeError exception if any errors were encountered during the backfilling.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py (hunk 18 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py]: It was rectified while addressing the comment below, the one about making failures more granular.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py (hunk 21 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:123]: I will add comments which hopefully make the code easier to follow.

However, I don't really see how path manipulations may be simplified if we would like to keep current flexibility in specification of the file sources.

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst (hunk 20 above)
Comment (mxk62) [reply doc/lsst.dbb.buffmngrs.endpoint/get-started.rst]: I will change the wording in the documentation and add a guard against absolute paths.

File: python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py (hunk 13 above)
Comment (mxk62) [reply python/lsst/dbb/buffmngrs/endpoint/backfill/commands.py:51]: I'll factor our repeated code to functions.

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 24]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+          file:
+            table: files
+          event:
+            table: gen2_file_events
+      backfill:
+        storage: /data/storage
+        sources:
+          - "2020-12-31"
+        search:
+          blacklist:
```
Comment (ktlim) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:312]: Please use a term like `exclude` rather than `blacklist`.

File: doc/lsst.dbb.buffmngrs.endpoint/get-started.rst [hunk 25]
```diff
@@ -269,13 +269,80 @@ to make another ingest attempt for selected files.
...
+
+      You can specify multiple sources.  The source entry can be either a file
+      or a directory.  You can use unix style pathname pattern expansion
+      when specifying source's name (tilde expansion is not supported though).
+      All sources must be relative with regard to the storage area.
+
+   .. note::
+
+      To prevent Python YAML parser from accidental conversion of strings like
+      ``2021-01-01`` into a ``datetime`` objects, enclose them in quotes.
```
Comment (ktlim) [inline doc/lsst.dbb.buffmngrs.endpoint/get-started.rst:325]: "into a" → "into"

File: python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py [hunk 26]
```diff
@@ -0,0 +1,224 @@
...
+        required = {"event", "file"}
+        missing = required - set(config["tablenames"])
+        if missing:
+            msg = f"invalid ORMs: {', '.join(missing)} not provided"
+            logger.error(msg)
+            raise ValueError(msg)
+        self.Event = event_creator(config["tablenames"])
+        self.File = file_creator(config["tablenames"])
+
+        self.storage = os.path.abspath(config["storage"])
```
Comment (ktlim) [inline python/lsst/dbb/buffmngrs/endpoint/backfill/backfill.py:87]: If you are doing lots of `os.path` calls, it may be better to use `pathlib`.

File: python/lsst/dbb/buffmngrs/endpoint/search.py [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
```
Comment (ktlim) [inline python/lsst/dbb/buffmngrs/endpoint/search.py:60]: Please change the name of the argument throughout this module.

**Review by ktlim — COMMENTED** [review]
Apologies for not looking at this earlier, but I have a few suggestions.

**Comment by mxk62** [comment]
If you don't mind, I'd like to address the issues you raised as a part of the other ticket I'm currently working on (DM-28494).

**Comment by ktlim** [comment]
Yes, that's fine.
